domain/port/state/state.go

```diff
@@ -185,28 +228,22 @@ func (st *State) UpdateUnitPorts(
 	unitUUID := unitUUID{UUID: unit}
 
 	return domain.Run(ctx, func(ctx context.Context, tx *sqlair.TX) error {
-		currentOpenedPorts, err := st.getUnitOpenedPorts(ctx, tx, unitUUID)
-		if err != nil {
-			return errors.Annotatef(err, "getting opened ports for unit %q", unit)
-		}
-
-		allInputPortRanges := append(openPorts.UniquePortRanges(), closePorts.UniquePortRanges()...)
-		err = verifyNoPortRangeConflicts(currentOpenedPorts, allInputPortRanges)
+		endpoints, err := st.ensureEndpoints(ctx, tx, unitUUID, endpointsUnderAction)
```

Comment:
If you describe what this method does in plain English, it should be obvious what should be lifted into the service layer as orchestration, and what should remain in state as modules of retrieval/persistence.